### ISSUE TITLE
Keep the started language servers without shutting down when the last editor is closed

### ISSUE THREAD

**Issue by gayanper** [spring-projects/sts4#200]
Today STS4 start the lang server when the first java class is opened, this adds a cpu overhead of starting up the lang server when opening the first editor which some times cause lagging through out eclipse. At the same time closing the last editor cause the close operation to hang sometime as well if the language server is still starting in progress.

Can we start language server if there is a java project or a project with spring (sts4) nature so that lang server is started up just after finishing workspace init in background. This could enhance the responsiveness of eclipse and reuse the started server until the workspace is closed.

**Comment by martinlippert** [comment]
This behavior is implemented in the overall LSP support for Eclipse (in the lsp4e project) and is the same across all language servers in Eclipse. We can think about changing that in order to startup the language server earlier (during workbench startup, for example), but that would cause the CPU consumption to happen when the IDE starts, which might not create a much better impression of the overall IDE performance.

Since all the language server work is done in a separate JVM process, it should not interfere with your IDE experience a lot and do things in parallel. So the question here is: what causes the "lagging" that you describe? Is it the IDE waiting for the language server to do something or to reply to a request (or something like that), or is it the CPU being busy doing language server work and the OS is not being able to shift enough CPU resources to the IDE process?

Depending on the outcome of this analysis we can start to think about further optimizations in this area... :-)

**Comment by martinlippert** [comment]
The latest CI builds include new improvement that caches symbols for projects, so re-starts of the language server should be much less expensive than in the past. Would be interesting to hear about your experiences using the latest CI builds (and the cache) and to learn whether that makes a difference for your request to keep the language server running.

**Comment by martinlippert** [comment]
This is in place now since we leverage the option of LSP4E to keep the language server running for a certain period of time after closing the last editor. This avoids the language server being shutdown and restarted all the time when you close all editors and open a new one to work on another file. However, the language server will be shutdown after a specific timeout, so it doesn't keep running forever.

**Comment by martinlippert** [comment]
Closed as duplicate of https://github.com/spring-projects/sts4/issues/568
